### DISCUSSION TITLE
"asking" micropython to run a script written on the device from the REPL or the WebREPL

### DISCUSSION THREAD

**Discussion by bonne-soeur** [micropython/micropython#10145] [Using MicroPython] [answered]
How to simply run a micropython script which is recorded on the device?
I would like to launch it from the REPL, and also when interacting through the WebREPL html page.

Of course when it is available and executable on the LOCAL HOST  environment, one can use "mpremote".
"mpremote" is not a command executred on the device as far as I understand
And I dont see how to use it when I am on the WebREPL page? 
Is there any command that the device interpreter takes in charge by itself to run a local script?
typing the name of the script like a command does not work.

Some people do copy the toto.py script as the main.py and reboot.....
but then it will start at any reboot and the interaction through the webrepl will get in trouble

Some other people propose to "import" the script as it would be a library. Then , as a feature of python library, any code outside any function declaration will be executed at load time, ONCE. Re-importing will not do any thing!

So what I lack is the equivalent of having a local "python" command stating that the next parameter on the line is the name of the file containing things to be interpreted. Or adding a directive at the first line of the script to tell that it needs to be in an environment with python interpreter. But here, the interpreter is always there, and does not need to be called, we are in the micropython REPL! . aren't we? 
Once you "give" something to the REPL, why does the interpreter does not act like the unix-shell in searching to identify the command  in a search-tree containing at last the local directory? 

So I might have missed something.The solution I found, in order to only use Micropython interpreter is to use exec() , like in the next sequence, in order to run 'toto.py':
>>> fid1 = open('toto.py')
>>> toto_str = fid1.read()
>>> fid1.close()
>>> exec(toto_str)
I used it only for short script. I have not tried it on a big loopy script. What else to do?
Thanks

**Comment by Carglglz** [comment, marked as answer]
@bonne-soeur 
>  Some other people propose to "import" the script as it would be a library. Then , as a feature of python library, any code outside any function declaration will be executed at load time, ONCE. Re-importing will not do any thing!

To "reload" / be able to run it again at REPL/WebREPL:

```
>>> import toto
....
>>> import sys
>>> del(sys.modules['toto'])
>>> import toto
....
```

`import` will check `sys.modules` dict and if `toto` is there it won't load it, so basically just delete it from there and it can be imported again 👍🏼 .

**Reply by bonne-soeur** [reply]
Hello Carglglz
Yes , it is a good point to remind here how to discharge the 'toto' library by typing:
>>> import sys
>>> del(sys.modules['toto'])

and then, it is understood that you can re-import it.You are 100% right.

But have you ever seen a python implementation from within which you can not run a python script by a command such
>>> python3 toto.py
?
Furthermore, on the github Mycropython repository, one can directly read from the readme file :
"MicroPython can execute scripts in textual source form (.py files) or from precompiled bytecode (.mpy files), in both cases either from an on-device filesystem or "frozen" into the MicroPython executable."
We are just discussing about executing  textual .py files which are sitting on the device file system, aren't we?
Is it a specific problem of the ESP8266 implementation? I have to check what I get on the RP2 too.

I would like to try to find in the sources where all the embedded instructions and libraries are declared and coded, 
and then I would include a new one called "mp" and code it directly , for instance in python , by using the lines that I proposed before:
def mp(script_str):
    fid1 = open(script_str)
    toto_str = fid1.read()
    fid1.close()
    exec(toto_str)
The syntax will be different from the original "python3 toto.py" as it will become pm('toto.py') 
and as far as python3 is concerned , we like function syntax, don't we?
Best Regards

**Reply by bonne-soeur** [reply]
Sorry, I finally meant:
>>> mp('toto.py')

**Comment by bonne-soeur** [comment]
So I went one step ahead and i tried to write a python library (= module) with the definition of mp() function:

```py
# library basic version
# exec parameter is the whole script text in one string

def mp(script_name):
    fid1=open(script_name)
    toto=fid1.read()
    fid1.close()
    exec(toto)        

print('the basic mp_lib is imported')
#that's all , folks
```

calling mp() on simple script works fine, 
one would do:

```
>>> from mp_lib import mp
>>> mp('little_script.py')
```

it works fine, even por several levels of loop in the program

but as soon as there are  internal function in the script to be excuted,
then there is a scope problem: the function cannot reach names that are defined outside their definitions!

basic example1:

```py
#test_scope1.py
def f1(x):
    y=x+foo
    return(y)
foo=4
a1=f1(3)
a2=f1(a1)
print(a2)
```

```
>>> mp('test_scope1.py')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\yvesb\local_python\mp_lib.py", line 8, in mp
    exec(toto)
  File "<string>", line 6, in <module>
  File "<string>", line 3, in f1
NameError: name 'foo' is not defined
>>> 
```

as one can read "name 'foo' is not defined.
Of course then, if one defines foo within the function definition, everything works fine:

```py
#test_scope3.py
def f1(x):
    foo=6
    y=x+foo
    return(y)
# foo=4
a1=f1(3)
a2=f1(a1)
print(a2)
```

```
>>> mp('test_scope3.py')
15
>>> 
```

This is very annoying. For instance if you import the machine library within your code, you can not use it within a function, or you need to import it inside the definition, and it will be (partly) executed at evey call!

any idea?

**Comment by bonne-soeur** [comment]
Sorry for the bad indentations for the scripts! 
I have not managed to edit properly my message. I shall improve

```py
# library basic version
# exec parameter is the whole script text in one string
def mp(script_name):
        fid1=open(script_name)
        toto=fid1.read()
        fid1.close()
        exec(toto)        
print('the basic mp_lib is imported')
#that's all , folks
```

**Reply by jimmo** [reply]
@bonne-soeur please see the [Welcome Post](https://github.com/orgs/micropython/discussions/9111) for instructions.

**Reply by bonne-soeur** [reply]
Merci jimmo ```py print('I shall improve evry day')```
